### PR TITLE
[ao] Added Equalization QConfig generation to ModelReport class

### PR DESCRIPTION
**Pull request by vspenubarthi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83698

Summary: This adds the capability to generate a QConfigMapping based on
the suggestions of the ModelReport API for the user to use. The only
dependency of this feature is that the calibration is run before the
generation of the QConfigMapping and there is no dependency on the
report generation other than that the observers cannot be removed before
this is called. This maps module fqns to EqualizationQConfigs instead of regular
QConfigs.

Example Usage (after callibration):

```
quantization_mapping = mod_report.generate_qconfig_mapping()
equalization_mapping = mod_report.generate_equalization_mapping()

prepared_model = quantize_fx.prepare_fx(model, mapping, example_input, _equalization_config=equalization_mapping)

quantized_model = quantize_fx.convert_fx(prepared)
```

This was tested by ensuring that the suggestions generated in the QConfigMapping are:
	1.	Correct according to the set backend and data passed through
	2.	Able to be prepared and converted as a proper config (is a valid config)
The test for this is a part of the TestFxModelReportClass test suite.

Test Plan: python test/test_quantization.py TestFxModelReportClass.test_equalization_mapping_generation

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/quantization/fx/_model_report/model_report.py [hunk 1]
```diff
@@ -490,20 +522,63 @@ def generate_qconfig_mapping(self) -> QConfigMapping:
...
+        module: torch.nn.Module
+    ) -> EqualizationQConfig:
+        r"""
+        We ignore the module argument here, and only focus on thedetector_qconfig_info
+
+        Returns the equalization configuration generated by the DetectorQConfigInfo object
+        """
+        return detector_qconfig_info.generate_equalization_qconfig()
+
     def generate_equalization_mapping(self) -> QConfigMapping:
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/model_report.py:580]: the name is equalization config I think: https://pytorch.org/docs/master/generated/torch.quantization.quantize_fx.prepare_fx.html#torch.quantization.quantize_fx.prepare_fx

Comment (vspenubarthi) [reply torch/ao/quantization/fx/_model_report/model_report.py:580]: that's for the config file, but this is for the mapping that maps each module fqn to config

File: torch/ao/quantization/fx/_model_report/model_report.py [hunk 2]
```diff
@@ -517,4 +588,16 @@ def generate_equalization_mapping(self) -> QConfigMapping:
 
         Returns a QConfigMapping for the equalization configuration
         """
-        pass
+        # get the mapping info
+        detector_qconfig_info_compiled = self._generate_qconfig_mapping(
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/model_report.py:592]: nit: do you mean combined here?

Comment (jerryzh168) [reply torch/ao/quantization/fx/_model_report/model_report.py:592]: same for the previous function

File: torch/ao/quantization/fx/_model_report/model_report.py [hunk 3]
```diff
@@ -461,7 +487,11 @@ def generate_qconfig_mapping(self) -> QConfigMapping:
         These configs are based on the suggestions provided by the ModelReport API
         and can only be generated once the reports have been generated.
 
-        Returns a QConfigMapping for the quantization configuration
+        Args:
+            update_qconfig_info_function (Callable) takes in a function that takes in two DetectorQConfigInfo
+            and updates the one that is being compiled
+
+        Returns a Dict mapping module_fqns to DetectorQConfigInfo objects
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/model_report.py:494]: this does not look like a qconfig_mapping, should we change the name?

Comment (vspenubarthi) [reply torch/ao/quantization/fx/_model_report/model_report.py:494]: changed it to `_generate_module_fqn_to_detector_info_mapping`

**Review by jerryzh168 — APPROVED** [review]
looks good

File: torch/ao/quantization/fx/_model_report/model_report.py [hunk 4]
```diff
@@ -485,27 +518,68 @@ def generate_qconfig_mapping(self) -> QConfigMapping:
...
-        mapping: QConfigMapping = self._generate_qconfig_mapping_helper(detector_qconfig_info_compiled)
+        mapping: QConfigMapping = self._generate_qconfig_mapping_helper(
+            detector_qconfig_info_combined,
+            self._quantization_config_generator
+        )
 
         # return the generated mapping
         return mapping
 
+    def _quantization_config_generator(self, detector_qconfig_info: DetectorQConfigInfo, module: torch.nn.Module) -> QConfig:
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/model_report.py:565]: I feel ideally we'd like to remove these one-line functions

Comment (jerryzh168) [reply torch/ao/quantization/fx/_model_report/model_report.py:565]: creating too many helper functions will also make the code less readable

**Pull request closed without merging**